File: CallX2/app/src/main/java/com/callx/app/adapters/GroupMemberAdapter.java
package com.callx.app.adapters;

import android.content.Context;
import android.view.*;
import android.widget.*;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.recyclerview.widget.RecyclerView;
import com.bumptech.glide.Glide;
import com.callx.app.R;
import com.callx.app.utils.Constants;
import java.text.SimpleDateFormat;
import java.util.*;

/**
 * GroupMemberAdapter — Ultra-advanced member list for GroupInfoActivity.
 *
 * Shows: circular avatar, online dot, name, admin/creator badge,
 *        last-seen text, 3-dot options menu (view profile, message,
 *        make admin, remove — admin-gated).
 */
public class GroupMemberAdapter extends RecyclerView.Adapter<GroupMemberAdapter.VH> {

    public interface OnMemberActionListener {
        void onAction(String uid, String action);
    }

    public static class MemberItem {
        public final String uid;
        public final String name;
        public final String role;   // "admin" | "member" | "creator"
        public final String photoUrl;
        public final boolean online;
        public final Long   lastSeen;

        public MemberItem(String uid, String name, String role,
                          String photoUrl, boolean online, Long lastSeen) {
            this.uid      = uid;
            this.name     = name;
            this.role     = role;
            this.photoUrl = photoUrl;
            this.online   = online;
            this.lastSeen = lastSeen;
        }
    }

    private final List<MemberItem>   items;
    private final String             currentUid;
    private final OnMemberActionListener listener;
    private boolean isAdmin = false;

    public GroupMemberAdapter(List<MemberItem> items, String currentUid,
                              OnMemberActionListener listener) {
        this.items      = items;
        this.currentUid = currentUid;
        this.listener   = listener;
    }

    public void setIsAdmin(boolean admin) {
        this.isAdmin = admin;
        notifyDataSetChanged();
    }

    @NonNull @Override
    public VH onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View v = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.item_group_member, parent, false);
        return new VH(v);
    }

    @Override
    public void onBindViewHolder(@NonNull VH h, int pos) {
        MemberItem m   = items.get(pos);
        Context ctx    = h.itemView.getContext();
        boolean isMe   = currentUid.equals(m.uid);

        // Name — append "(You)" for self
        h.tvName.setText(isMe ? m.name + " (You)" : m.name);

        // Role badge
        boolean showBadge = "admin".equals(m.role) || "creator".equals(m.role);
        h.tvRoleBadge.setVisibility(showBadge ? View.VISIBLE : View.GONE);
        if (showBadge) h.tvRoleBadge.setText("creator".equals(m.role) ? "Creator" : "Admin");

        // Online dot
        h.onlineDot.setVisibility(m.online ? View.VISIBLE : View.GONE);

        // Avatar
        if (m.photoUrl != null && !m.photoUrl.isEmpty()) {
            Glide.with(ctx).load(m.photoUrl)
                    .placeholder(R.drawable.ic_person)
                    .circleCrop()
                    .into(h.ivAvatar);
        } else {
            h.ivAvatar.setImageResource(R.drawable.ic_person);
        }

        // Last seen / status
        if (m.online) {
            h.tvStatus.setText("Online");
        } else if (m.lastSeen != null && m.lastSeen > 0) {
            h.tvStatus.setText("last seen " + formatLastSeen(m.lastSeen));
        } else {
            h.tvStatus.setText("");
        }

        // Options menu
        h.btnOptions.setOnClickListener(v -> {
            if (isMe) {
                // My own row: only profile info
                listener.onAction(m.uid, "view_profile");
                return;
            }
            showMemberOptionsMenu(ctx, m);
        });

        // Row click = view profile
        h.itemView.setOnClickListener(v -> listener.onAction(m.uid, "view_profile"));
    }

    private void showMemberOptionsMenu(Context ctx, MemberItem m) {
        List<String>  labels  = new ArrayList<>();
        List<String>  actions = new ArrayList<>();

        labels.add("Message"); actions.add("message");
        labels.add("View Profile"); actions.add("view_profile");

        if (isAdmin && !"creator".equals(m.role)) {
            if ("admin".equals(m.role)) {
                labels.add("Revoke Admin"); actions.add("revoke_admin");
            } else {
                labels.add("Make Admin 👑"); actions.add("make_admin");
            }
            labels.add("Remove from Group"); actions.add("remove");
        }

        new AlertDialog.Builder(ctx)
                .setTitle(m.name)
                .setItems(labels.toArray(new String[0]), (d, which) ->
                        listener.onAction(m.uid, actions.get(which)))
                .show();
    }

    @Override public int getItemCount() { return items.size(); }

    private static String formatLastSeen(long ts) {
        long diff = System.currentTimeMillis() - ts;
        if (diff < 60_000) return "just now";
        if (diff < 3_600_000) return (diff / 60_000) + " min ago";
        if (diff < 86_400_000) return (diff / 3_600_000) + " hours ago";
        SimpleDateFormat sdf = new SimpleDateFormat("dd MMM", Locale.getDefault());
        return sdf.format(new Date(ts));
    }

    static class VH extends RecyclerView.ViewHolder {
        de.hdodenhof.circleimageview.CircleImageView ivAvatar;
        View   onlineDot;
        TextView tvName, tvRoleBadge, tvStatus;
        View   btnOptions;

        VH(@NonNull View itemView) {
            super(itemView);
            ivAvatar    = itemView.findViewById(R.id.iv_avatar);
            onlineDot   = itemView.findViewById(R.id.online_dot);
            tvName      = itemView.findViewById(R.id.tv_member_name);
            tvRoleBadge = itemView.findViewById(R.id.tv_role_badge);
            tvStatus    = itemView.findViewById(R.id.tv_member_status);
            btnOptions  = itemView.findViewById(R.id.btn_member_options);
        }
    }
}

File: CallX2/feature-chat/src/main/java/com/callx/app/adapters/GroupMemberAdapter.kt
package com.callx.app.adapters

import android.view.*
import android.widget.*
import androidx.appcompat.app.AlertDialog
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.callx.app.chat.R
import com.callx.app.utils.Constants
import java.text.SimpleDateFormat
import java.util.*

/**
 * GroupMemberAdapter — Member list for GroupInfoActivity.
 */
class GroupMemberAdapter(
    private val items: MutableList<MemberItem>,
    private val currentUid: String,
    private val listener: OnMemberActionListener
) : RecyclerView.Adapter<GroupMemberAdapter.VH>() {

    interface OnMemberActionListener { fun onAction(uid: String, action: String) }

    data class MemberItem(
        val uid:      String,
        val name:     String,
        val role:     String,
        val photoUrl: String?,
        val online:   Boolean,
        val lastSeen: Long?
    )

    var isAdmin = false
    private val dateFmt = SimpleDateFormat("MMM dd", Locale.getDefault())

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): VH {
        val v = LayoutInflater.from(parent.context).inflate(R.layout.item_group_member, parent, false)
        return VH(v)
    }

    override fun onBindViewHolder(h: VH, pos: Int) {
        val item = items[pos]
        h.tvName.text  = item.name
        h.tvRole.text  = when (item.role) { "admin" -> "Admin 👑"; "creator" -> "Creator"; else -> "" }
        h.tvRole.visibility = if (item.role == "member") View.GONE else View.VISIBLE
        h.onlineDot.visibility = if (item.online) View.VISIBLE else View.GONE

        if (!item.photoUrl.isNullOrEmpty()) {
            Glide.with(h.itemView.context).load(item.photoUrl).circleCrop()
                .placeholder(R.drawable.ic_person).into(h.ivAvatar)
        } else h.ivAvatar.setImageResource(R.drawable.ic_person)

        val lastSeenText = when {
            item.online  -> "Online"
            item.lastSeen != null -> "Last seen ${dateFmt.format(Date(item.lastSeen))}"
            else -> ""
        }
        h.tvLastSeen.text = lastSeenText

        h.btnMore.visibility = if (isAdmin && item.uid != currentUid) View.VISIBLE else View.GONE
        h.btnMore.setOnClickListener { showOptions(h.itemView, item) }
    }

    private fun showOptions(anchor: View, item: MemberItem) {
        val isAdm = item.role == "admin"
        val opts = arrayOf("View Profile", "Message", if (isAdm) "Revoke Admin" else "Make Admin", "Remove from Group")
        AlertDialog.Builder(anchor.context)
            .setTitle(item.name)
            .setItems(opts) { _, which ->
                val action = when (which) { 0 -> "view"; 1 -> "message"; 2 -> if (isAdm) "revoke_admin" else "make_admin"; else -> "remove" }
                listener.onAction(item.uid, action)
            }.show()
    }

    override fun getItemCount() = items.size

    class VH(v: View) : RecyclerView.ViewHolder(v) {
        val tvName:    TextView         = v.findViewById(R.id.tv_member_name)
        val tvRole:    TextView         = v.findViewById(R.id.tv_member_role)
        val tvLastSeen: TextView        = v.findViewById(R.id.tv_last_seen)
        val ivAvatar:  de.hdodenhof.circleimageview.CircleImageView = v.findViewById(R.id.iv_avatar)
        val onlineDot: View             = v.findViewById(R.id.view_online_dot)
        val btnMore:   ImageButton      = v.findViewById(R.id.btn_more)
    }
}
